### ISSUE TITLE
Test failures with PDL 2.091

### ISSUE THREAD

**Issue by sebastic** [moocow-the-bovine/PDL-VectorValued#10]
Several tests fail with PDL 2.091:
```
 30s autopkgtest [22:12:03]: test autodep8-perl-build-deps: [-----------------------
 31s t/00_basic.t ... 
 31s 1..2
 31s ok 1 - use PDL;
 31s ok 2 - use PDL::VectorValued;
 31s ok
 31s t/01_rlevec.t .. 
 31s 1..17
 31s loaded /tmp/autopkgtest-lxc.fhbgv833/downtmp/autopkgtest_tmp/smokeCjhUE3/t/common.plt
 31s ok 1 - rlevec():counts
 31s ok 2 - rlevec():elts
 31s ok 3 - rldvec()
 31s ok 4 - enumvec()
 31s ok 5 - enumvecg()
 31s ok 6 - rleND():2d:counts
 31s ok 7 - rleND():2d:elts
 31s ok 8 - rldND():2d
 31s ok 9 - rleND():Nd:counts
 31s ok 10 - rleND():Nd:elts
 31s ok 11 - rldND():Nd
 31s ok 12 - enumvec():Nd
 31s ok 13 - rldseq():type
 31s ok 14 - rldseq():data
 31s ok 15 - rleseq():type
 31s ok 16 - rleseq():lens
 31s ok 17 - rleseq():offs
 31s 
 31s ok
 31s t/02_cmpvec.t .. 
 31s 1..8
 31s loaded /tmp/autopkgtest-lxc.fhbgv833/downtmp/autopkgtest_tmp/smokeCjhUE3/t/common.plt
 31s ok 1 - cmpvec:1d:<
 31s ok 2 - cmpvec:1d:>
 31s ok 3 - cmpvec:1d:==
 31s ok 4 - vv_qsortvec
 31s ok 5 - vv_qsortveci
 31s ok 6 - vsearchvec():match
 31s ok 7 - vsearchvev():<<
 31s ok 8 - vsearchvev():>>
 31s 
 31s ok
 31s t/03_setops.t .. 
 31s loaded /tmp/autopkgtest-lxc.fhbgv833/downtmp/autopkgtest_tmp/smokeCjhUE3/t/common.plt
 31s ok 1 - vv_union:list:c
 31s ok 2 - vv_union:list:nc
 31s ok 3 - vv_union:scalar
 31s ok 4 - vv_intersect:list:c
 31s ok 5 - vv_intersect:list:nc
 31s ok 6 - vv_intersect:scalar
 31s ok 7 - vv_setdiff:list:c
 31s ok 8 - vv_setdiff:list:nc
 31s ok 9 - vv_setdiff:scalar
 31s ok 10 - v_union
 31s ok 11 - v_intersect
 31s ok 12 - v_setdiff
 31s ok 13 - vv_union - broadcast dims - uw+wx
 31s ok 14 - vv_union - broadcast dims - uw+xy
 31s ok 15 - vv_union - broadcast dims - 0+wx
 31s ok 16 - vv_union - broadcast dims - wx+0
 31s ok 17 - vv_union - broadcast dims - 0+0
 31s ok 18 - vv_union - broadcast dims - uw(*k)+wx
 31s ok 19 - vv_union - broadcast dims - uw(*k)+xy
 31s ok 20 - vv_union - broadcast dims - 0(*k)+wx
 31s ok 21 - vv_union - broadcast dims - wx(*k)+0
 31s ok 22 - vv_union - broadcast dims - 0(*k)+0
 31s ok 23 - vv_intersect - broadcast dims - needle0&haystack
 31s ok 24 - vv_intersect - broadcast dims - needle1&haystack
 31s ok 25 - vv_intersect - broadcast dims - needles&haystack
 31s ok 26 - vv_intersect - broadcast dims - haystack&haystack
 31s ok 27 - vv_intersect - broadcast dims - haystack&empty
 31s ok 28 - vv_intersect - broadcast dims - empty&haystack
 31s ok 29 - vv_intersect - broadcast dims - needle0(*k)&haystack
 31s ok 30 - vv_intersect - broadcast dims - needle1(*k)&haystack
 31s ok 31 - vv_intersect - broadcast dims - needles(*k)&haystack
 31s ok 32 - vv_intersect - broadcast dims - haystack(*k)&haystack
 31s ok 33 - vv_intersect - broadcast dims - haystack(*k)&empty
 31s ok 34 - vv_intersect - broadcast dims - empty(*k)&haystack
 31s ok 35 - vv_setdiff - broadcast dims - haystack-needle0
 31s ok 36 - vv_setdiff - broadcast dims - haystack-needle1
 31s ok 37 - vv_setdiff - broadcast dims - haystack-needles
 31s ok 38 - vv_setdiff - broadcast dims - haystack-haystack
 31s ok 39 - vv_setdiff - broadcast dims - haystack-empty
 31s ok 40 - vv_setdiff - broadcast dims - empty-haystack
 31s ok 41 - vv_setdiff - broadcast dims - haystack(*k)-needle0
 31s ok 42 - vv_setdiff - broadcast dims - haystack(*k)-needle1
 31s ok 43 - vv_setdiff - broadcast dims - haystack(*k)-needles
 31s ok 44 - vv_setdiff - broadcast dims - haystack(*k)-haystack
 31s ok 45 - vv_setdiff - broadcast dims - haystack(*k)-empty
 31s ok 46 - vv_setdiff - broadcast dims - empty(*k)-haystack
 31s ok 47 - vv_intersect - implicit dims - titi&toto
 31s ok 48 - vv_intersect - implicit dims - notin&toto
 31s ok 49 - vv_intersect - implicit dims - titi(*1)&toto
 31s ok 50 - vv_intersect - implicit dims - notin(*1)&toto
 31s ok 51 - vv_intersect - implicit dims - needle0_in&haystack - result
 31s ok 52 - vv_intersect - implicit dims - needle0_in&haystack - counts
 31s ok 53 - vv_intersect - implicit dims - needle0_in&haystack - scalar
 31s ok 54 - vv_intersect - implicit dims - needle_in&haystack - result
 31s ok 55 - vv_intersect - implicit dims - needle_in&haystack - counts
 31s ok 56 - vv_intersect - implicit dims - needle_in&haystack - scalar
 31s ok 57 - vv_intersect - implicit dims - needle0_notin&haystack - result
 31s ok 58 - vv_intersect - implicit dims - needle0_notin&haystack - counts
 31s ok 59 - vv_intersect - implicit dims - needle0_notin&haystack - scalar
 31s ok 60 - vv_intersect - implicit dims - needle_notin&haystack - result
 31s ok 61 - vv_intersect - implicit dims - needle_notin&haystack - counts
 31s ok 62 - vv_intersect - implicit dims - needle_notin&haystack - scalar
 31s ok 63 - vv_intersect - implicit dims - needles&haystack - result
 31s ok 64 - vv_intersect - implicit dims - needles&haystack - counts
 31s ok 65 - vv_intersect - implicit dims - needles&haystack - scalar
 31s ok 66 - vv_intersect - implicit dims - needles8x&haystack8 - result
 31s ok 67 - vv_intersect - implicit dims - needles8x&haystack8 - counts
 31s ok 68 - vv_intersect - implicit dims - needles8x&haystack8 - scalar
 31s ok 69 - v_union - broadcast dims - 12+34
 31s ok 70 - v_union - broadcast dims - 34+1234
 31s ok 71 - v_union - broadcast dims - 0+1234
 31s ok 72 - v_union - broadcast dims - 1234+0
 31s ok 73 - v_union - broadcast dims - 0+0
 31s ok 74 - v_union - broadcast dims - 12(*k)+34
 31s ok 75 - v_union - broadcast dims - 34(*k)+1234
 31s ok 76 - v_union - broadcast dims - 0(*k)+1234
 31s ok 77 - v_union - broadcast dims - 1234(*k)+0
 31s ok 78 - v_union - broadcast dims - 0(*k)+0
 31s ok 79 - v_intersect - broadcast dims - 12&34
 31s ok 80 - v_intersect - broadcast dims - 34&1234
 31s ok 81 - v_intersect - broadcast dims - 0&1234
 31s ok 82 - v_intersect - broadcast dims - 1234&0
 31s ok 83 - v_intersect - broadcast dims - 0&0
 31s ok 84 - v_intersect - broadcast dims - 12(*k)&34
 31s ok 85 - v_intersect - broadcast dims - 34(*k)&1234
 31s ok 86 - v_intersect - broadcast dims - 0(*k)&1234
 31s ok 87 - v_intersect - broadcast dims - 1234(*k)&0
 31s ok 88 - v_intersect - broadcast dims - 0(*k)&0
 31s ok 89 - v_setdiff - broadcast dims - 12-34
 31s ok 90 - v_setdiff - broadcast dims - 34-1234
 31s ok 91 - v_setdiff - broadcast dims - 1234-0
 31s ok 92 - v_setdiff - broadcast dims - 0-1234
 31s ok 93 - v_setdiff - broadcast dims - 0-0
 31s ok 94 - v_setdiff - broadcast dims - 12(*k)-34
 31s ok 95 - v_setdiff - broadcast dims - 34(*k)-1234
 31s ok 96 - v_setdiff - broadcast dims - 1234(*k)-0
 31s ok 97 - v_setdiff - broadcast dims - 0(*k)-1234
 31s ok 98 - v_setdiff - broadcast dims - 0(*k)-0
 31s 
 31s 1..98
 31s ok
 31s t/04_types.t ... 
 31s 1..2
 31s loaded /tmp/autopkgtest-lxc.fhbgv833/downtmp/autopkgtest_tmp/smokeCjhUE3/t/common.plt
 31s ok 1 - isa(vv_indx,PDL::Type)
 31s ok 2 - vv_indx == PDL::indx (PDL >= v2.007)
 31s 
 31s ok
 31s t/05_vcos.t .... 
 31s 1..6
 31s loaded /tmp/autopkgtest-lxc.fhbgv833/downtmp/autopkgtest_tmp/smokeCjhUE3/t/common.plt
 31s ok 1 - vv_vcos:flat
 31s ok 2 - vv_vcos:broadcasted
 31s ok 3 - vv_vcos:nullvec:a:nan
 31s ok 4 - vv_vcos:nullvec:b:all-nan
 31s not ok 5 - vv_vcos:bad:a
 31s 
 31s #   Failed test 'vv_vcos:bad:a'
 31s #   at /tmp/autopkgtest-lxc.fhbgv833/downtmp/autopkgtest_tmp/smokeCjhUE3/t/common.plt line 42.
 31s # got=[NaN NaN NaN]
 31s # want=[0.93094 0.64549 -0.93094]
 31s not ok 6 - vv_vcos:bad:b
 31s 
 31s #   Failed test 'vv_vcos:bad:b'
 31s #   at /tmp/autopkgtest-lxc.fhbgv833/downtmp/autopkgtest_tmp/smokeCjhUE3/t/common.plt line 42.
 31s # got=[NaN NaN NaN]
 31s # want=[0.8366 0.6211 -0.8366]
 31s 
 31s # Looks like you failed 2 tests of 6.
 31s Dubious, test returned 2 (wstat 512, 0x200)
 31s Failed 2/6 subtests 
 31s 
 31s Test Summary Report
 31s -------------------
 31s t/05_vcos.t  (Wstat: 512 (exited 2) Tests: 6 Failed: 2)
 31s   Failed tests:  5-6
 31s   Non-zero exit status: 2
 31s Files=6, Tests=133,  1 wallclock secs ( 0.02 usr  0.01 sys +  0.60 cusr  0.15 csys =  0.78 CPU)
 31s Result: FAIL
 32s autopkgtest [22:12:05]: test autodep8-perl-build-deps: -----------------------]
```
[Full log](https://ci.debian.net/packages/libp/libpdl-vectorvalued-perl/testing/amd64/51301693/)

CC: @mohawk2

**Comment by moocow-the-bovine** [comment]
I am unable to reproduce this locally using current git HEADs:
- PDL::VectorValued 54adfda (v1.0.23 + irrelevant changes)
- PDL core [bcf018e](https://github.com/PDLPorters/pdl/commits/bcf018e) (v2.091)

@sebastic can you tell me where the PDL::VectorValued source for the deb build is coming from?  Are you cloning directly from github or fetching the distribution from CPAN?

In either case, it might be worth trying to re-run the tests after commenting out or deleting the `vv_vcos` entry in `%VV_IMPORT` at [VectorValued.pm line 53](https://github.com/moocow-the-bovine/PDL-VectorValued/blob/df2b32ddd76e9cc7f66d9fe6386069d122da3e79/VectorValued.pm#L53) (effectively reverting https://github.com/moocow-the-bovine/PDL-VectorValued/commit/df2b32d):
- without that line, the `vv_vcos()` function defined in this module is used
- with that line active, `vv_vcos()` is just an alias for `PDL::vcos` defined in PDL core (for VV v1.0.23 and PDL v2.091)

I tried both with & without the `vcos()` alias, but got no test failures in either case.

**Comment by sebastic** [comment]
> @sebastic can you tell me where the PDL::VectorValued source for the deb build is coming from? Are you cloning directly from github or fetching the distribution from CPAN?

From CPAN, see: [`debian/watch`](https://salsa.debian.org/perl-team/modules/packages/libpdl-vectorvalued-perl/-/blob/master/debian/watch?ref_type=heads#L2)

The test doesn't fail at build time with or without `vv_vcos()` in `%VV_IMPORT`, it does fail at runtime with and without  `vv_vcos()` in `%VV_IMPORT`.

**Comment by mohawk2** [comment]
I'm having trouble reproducing both this and https://github.com/PDLPorters/PDL-Stats/issues/33 - @sebastic do you have any ideas on how I can do so? Am I likely to need to do `chroot` type stuff on a proper Debian box? If so, is there a recipe for me to do so, starting with a new Vagrant Debian box?

**Comment by sebastic** [comment]
Create a cowbuilder chroot for sid if you haven't done so already:
```
sudo cowbuilder --create \
                --distribution=sid \
                --basepath=/var/cache/pbuilder/base-sid.cow
```
Update it if you already have the chroot:
```
sudo cowbuilder --update \
                --basepath=/var/cache/pbuilder/base-sid.cow
```
Then login to the chroot to run autopkgtest:
```
sudo cowbuilder --login --basepath /var/cache/pbuilder/base-sid.cow
# Inside chroot
cd /tmp/buildd
apt source libpdl-vectorvalued-perl
cd libpdl-vectorvalued-perl-*/
apt install newpid autopkgtest autodep8 apt-utils
newpid autopkgtest -B -- autopkgtest-virt-null
```

Alternatively you can use `sbuild` chroots as documented in:
 [autopkgtest - Debian Perl Group autopkgtest implementation and team practices](https://perl-team.pages.debian.net/autopkgtest.html#SETUP)

**Comment by mohawk2** [comment]
Results so far. I'm using this Vagrant setup:

```
Vagrant.configure("2") do |config|
  config.vm.box = "debian/testing64"
  config.vm.provider "virtualbox" do |v|
    v.memory = 4096
    v.cpus = 8
  end
end
```

The `cowbuilder` option hasn't gone great, but the `autopkgtest` option with modifications has at least reproduced the `vcos` failure, albeit bizarrely at the `autodep8-perl-build-deps` stage, using these commands (excerpted from https://perl-team.pages.debian.net/autopkgtest.html#SETUP) in a fresh Vagrant box:

```
sudo apt-get update
sudo apt-get install --install-recommends sbuild autopkgtest
sudo sbuild-createchroot sid /srv/sbuild/sid http://deb.debian.org/debian
sudo mv /etc/schroot/chroot.d/sid-amd64-sbuild-* /etc/schroot/chroot.d/sid-amd64-sbuild.conf
sudo sbuild-update -udr sid-amd64-sbuild
sudo autopkgtest libpdl-vectorvalued-perl -- schroot sid-amd64-sbuild
```

The various Debian group pages on these things are confusing and in some cases out of date, including messages from the commands; e.g. `sbuild-createchroot` confidently says on exit `I: Run "sbuild-adduser" to add new sbuild users.`, but that command does not exist.

I'm still working out how to get at the (s)chroot environment it creates to do further debugging, but I call this progress.

**Comment by moocow-the-bovine** [comment]
not sure if this is helpful, but fwiw I've also reproduced the failures using cowbuilder as described [above](https://github.com/moocow-the-bovine/PDL-VectorValued/issues/10#issuecomment-2335156159), and observed that:

- `vv_vcos` really does appear to be correctly resolving to `PDL::vcos`
- failures occur on invocation of `newpid autopkgtest -B -- autopkgtest-virt-null` (as above)
- failures do *not* occur without the `-B` option to `autopkgtest` (*"-B, --no-built-binaries : do not build/use binaries from .dsc, git source, or unbuilt tree"*)
- as soon as I've invoked `autopkgtest` once **without** the -B option, even subsequent runs of `autopkgtst` **with** the -B option succeed ... until I exit the cowbuilder chroot session
    - even if I `rm -rf /tmp/buildd/libpdl-vectorvalued*` and re-run the commands starting from `apt source libpdl-vectorvalued-perl`, `autopkgtest -B` succeeds after it's been run once in the chroot without `-B`

... this seems very weird to me, but I admit near-total ignorance of what autopkgtest is (or should be) doing here.  This might be expected behavior.  If so, sorry for the noise.

**Comment by mohawk2** [comment]
Something that very definitely does not help: `autopkgtest` builds into an `autopkgtest_tmp` directory, then after the test ran, **deletes the build directory before opening the requested shell**. That seems like a bug, and renders the shell a bit useless.

Also, @moocow-the-bovine, I did try the `cowbuilder` option first. Would you be able to make a list of starting-from-scratch instructions to go from nothing to being able to do the repro actions you were doing? I need to see the built files and poke at them.

**Comment by moocow-the-bovine** [comment]
> ... this seems very weird to me, but I admit near-total ignorance of what autopkgtest is (or should be) doing here. This might be expected behavior. If so, sorry for the noise.

this *does* seem to be expected behavior.  Without `-B`, autopktest appears to install the `libpdl-vectorvalued-perl` via `apt install`.  If I then `apt remove libpdl-vectorvalued-perl` in the chroot and re-run with `-B`, I get failures again.

**Comment by moocow-the-bovine** [comment]
> Would you be able to make a list of starting-from-scratch instructions to go from nothing to being able to do the repro actions you were doing? I need to see the built files and poke at them.

Starting from my local box running debian buster:
- run  the `sudo cowbuilder create` from [sebastic's comment](https://github.com/moocow-the-bovine/PDL-VectorValued/issues/10#issuecomment-2335156159)
- fire up the chroot as above: `sudo cowbuilder --login --basepath /var/cache/pbuilder/base-sid.cow`
- inside the chroot I needed some tiny tweaks:
```
perl -i -pe 's/^#//;' /etc/apt/sources.list && apt update # we want deb-src entries
apt install -y newpid autopkgtest autodep8 apt-utils
cd /tmp/buildd
apt source libpdl-vectorvalued-perl
cd libpdl-vectorvalued-perl-*/
newpid autopkgtest -B -- autopkgtest-virt-null
```

... except for the `sources.list` tweak, this is exactly as described in https://github.com/moocow-the-bovine/PDL-VectorValued/issues/10#issuecomment-2335156159 .

EDIT: I've tried using `autopkgtest -B -s` to get a shell after failed tests, but so far that hasn't been very enlightening.

**Comment by moocow-the-bovine** [comment]
could this problem be as easy (and as stupid) as a linking error?  With `-B -s` on a fresh chroot, I'm seeing 

```
...
autopkgtest [13:16:59]: @@@@@@@@@@@@@@@@@@@@ unbuilt-tree .
autopkgtest [13:16:59]: testing package libpdl-vectorvalued-perl version 1.0.23-1
autopkgtest [13:16:59]: build not needed
autopkgtest [13:16:59]: test autodep8-perl-build-deps: preparing testbed
...
Setting up pdl (1:2.091-1) ...
Setting up libpdl-vectorvalued-perl (1.0.23-1) ...
...
#   Failed test 'vv_vcos:bad:a'
...
Files=6, Tests=133,  1 wallclock secs ( 0.02 usr  0.03 sys +  0.82 cusr  0.05 csys =  0.92 CPU)
Result: FAIL
autopkgtest [13:17:11]: test autodep8-perl-build-deps: -----------------------]
autopkgtest [13:17:11]: test autodep8-perl-build-deps:  - - - - - - - - - - results - - - - - - - - - -
autodep8-perl-build-deps FAIL non-zero exit status 1
autopkgtest [13:17:11]:  - - - - - - - - - - running shell - - - - - - - - - -
```

Hypothesis:
- `autopkgtest` is installing a `libpdl-vectorvalued-perl` deb from an apt repo (log message `build not needed`)
- the VV deb was built with an older version of PDL
- some linker symbols are incompatible, and the VV `*.so`s need to be re-built against the new PDL

... that doesn't sound too likely either, because I'm pretty sure things would have broken before now if every new PDL release caused errors like this.  Not sure how the debian CI logic handles situations like that either.

@mohawk2 have there been any ABI changes for PDL-2.091 ?

**Comment by mohawk2** [comment]
The VV was built with an older PDL, and I didn't *think* there was any binary incompatibility. There *was* a change in getting badvalues from the core (required because of bugs in old clang), but my tests didn't show any binary breakage. @sebastic I think this might be fixable by (sigh) adding a patch to PDL reverting those binary changes. Does that sound plausible, and if so should I create those in the salsa repo?

I'll have a go at checking VV locally with before-and-after that API change to confirm.

**Comment by mohawk2** [comment]
Unfortunately, @moocow-the-bovine was entirely right. I'll put together a patch once I figure out the minimum.

**Comment by mohawk2** [comment]
@sebastic In fact, I've just released PDL 2.092 instead, since breaking ABI accidentally isn't a minor patching thing. If I'm right and that fixes this and https://github.com/PDLPorters/PDL-Stats/issues/33, can you let me know and/or close these? Else let me know and I'll try to fix it.

**Comment by moocow-the-bovine** [comment]
@mohawk2 if it would be any help, I'd be happy to bump the VV version & push out a new CPAN distribution ... presumably, the CI framework would build that against the current PDL in sid rather than trying to use the existing deb from buster.  Maybe the same technique could work for PDL::Stats too.  But that does seem like a rather painful way to deal with dependencies.

**Comment by mohawk2** [comment]
@moocow-the-bovine No, that wouldn't help. The fundamental problem here was that PDL broke ABI, so any other users of it (using Debian or otherwise) would have their modules break subtly until they were recompiled/installed. The fix for this is to have PDL be compatible with its old self, which it now is again, and Bas has already uploaded the packaged version. I assume soon enough we'll know whether that fixes it.

**Comment by sebastic** [comment]
The ABI break could also have been an argument to increment `PDL_CORE_VERSION` which triggers a transition for which all reverse dependencies will be rebuilt.

**Comment by mohawk2** [comment]
It could! But I don't want to cause the extra work for packaging heroes like yourself till I can't avoid it. If you're at all interested, you might see scattered through the core code the string `CORE21` which is stuff I will tidy up when I do update the ABI. There are a number of minor contortions I already have in there just to maintain compatibility.

I suspect that updating with new types, including possibly a bitfield type, will make it unavoidable to break ABI, but I'm not there yet.

**Comment by sebastic** [comment]
This is fixed with PDL 2.092, see: https://qa.debian.org/excuses.php?package=pdl

And the CI job: https://ci.debian.net/packages/libp/libpdl-vectorvalued-perl/testing/amd64/51323736/
